src/components/application_manager/src/policies/policy_handler.cc

```diff
@@ -1539,49 +1613,56 @@ void PolicyHandler::CheckPermissions(
 uint32_t PolicyHandler::GetNotificationsNumber(
     const std::string& priority) const {
   POLICY_LIB_CHECK(0);
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   return policy_manager_->GetNotificationsNumber(priority);
 }
 
 DeviceConsent PolicyHandler::GetUserConsentForDevice(
     const std::string& device_id) const {
   POLICY_LIB_CHECK(kDeviceDisallowed);
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   return policy_manager_->GetUserConsentForDevice(device_id);
 }
 
 bool PolicyHandler::GetDefaultHmi(const std::string& policy_app_id,
                                   std::string* default_hmi) const {
   POLICY_LIB_CHECK(false);
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   return policy_manager_->GetDefaultHmi(policy_app_id, default_hmi);
 }
 
 bool PolicyHandler::GetInitialAppData(const std::string& application_id,
                                       StringArray* nicknames,
                                       StringArray* app_hmi_types) {
   POLICY_LIB_CHECK(false);
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   return policy_manager_->GetInitialAppData(
       application_id, nicknames, app_hmi_types);
 }
 
 void PolicyHandler::GetUpdateUrls(const std::string& service_type,
                                   EndpointUrls& out_end_points) {
   POLICY_LIB_CHECK_VOID();
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   policy_manager_->GetUpdateUrls(service_type, out_end_points);
 }
 
 void PolicyHandler::GetUpdateUrls(const uint32_t service_type,
                                   EndpointUrls& out_end_points) {
   POLICY_LIB_CHECK_VOID();
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   policy_manager_->GetUpdateUrls(service_type, out_end_points);
 }
 
 std::string PolicyHandler::GetLockScreenIconUrl() const {
   POLICY_LIB_CHECK(std::string(""));
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   return policy_manager_->GetLockScreenIconUrl();
 }
 
 uint32_t PolicyHandler::NextRetryTimeout() {
   POLICY_LIB_CHECK(0);
-  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(policy_manager_lock_);
   return policy_manager_->NextRetryTimeout();
```

Comment:
@SNiukalov try to create similar template with the function return type

Reply:
@SNiukalov that's crazy, but something like that:
```
template <typename Func, typename... Args>
  auto CallPolicyManagerFunction(Func func, Args... args) -> decltype((std::declval<PolicyManager>().*std::declval<Func>())(std::declval<Args>()...)) {
    //sync_primitives::AutoLock lock(policy_manager_lock_);
    return ((*policy_manager_).*func)(args...);
  }
```

It should work for both cases - when return type required and when not

Reply:
@SNiukalov or another way - try to create accessor to PolicyManager instance and make access to it only through accessor

Reply:
@AKalinich-Luxoft 
 Fixed in: https://github.com/smartdevicelink/sdl_core/pull/2614/commits/793f178f719ad695595d1c8ad0ae40bcf265c9b5